src/update_handler.c

```diff
@@ -2852,21 +2776,24 @@ img_to_slot_handler get_update_handler(RaucImage *mfimage, RaucSlot *dest_slot,
 
 	g_message("Checking image type for slot type: %s", dest);
 
-	for (RaucUpdatePair *updatepair = updatepairs; updatepair->src != NULL; updatepair++) {
-		if (g_pattern_match_simple(updatepair->src, src) &&
-		    g_pattern_match_simple(updatepair->dest, dest)) {
-			g_message("Image detected as type: %s", updatepair->src);
-			handler = updatepair->handler;
-			break;
+	/* If no type is set, derive it from filename extension to support older version / migration case */
+	if (!mfimage->type) {
+		const gchar *derived_type = NULL;
+		if (derive_image_type_from_filename_pattern(src, &derived_type)) {
+			mfimage->type = g_strdup(derived_type);
+			g_message("Image detected as type: %s", mfimage->type);
 		}
 	}
 
-	if (handler == NULL) {
-		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_NO_HANDLER, "Unsupported image %s for slot type %s",
-				mfimage->filename, dest);
-		goto out;
+	if (mfimage->type) {
+		handler = get_handler_from_type(mfimage->type, dest);
+		if (handler) {
+			g_message("Found handler for image type '%s' and slot type '%s'", mfimage->type, dest);
+			return handler;
+		}
 	}
 
-out:
-	return handler;
+	g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_NO_HANDLER,
+			"Unsupported image %s for slot type %s", src, dest);
+	return NULL;
```

Comment:
Keep the good case as the normal path by checking the error case using `if (!mfimage->type)` or `if (!handler)`.